File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        stack<TreeNode*> s;
        TreeNode* curr = root;
        vector<int> res;
        while(curr || !s.empty())
        {
            if (curr) {
                res.push_back(curr->val);
                s.push(curr->right);
                curr = curr->left;
            } else {
                curr = s.top();
                s.pop();
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/validate-parentheses/submission-4.cs
public class Solution {
    public bool IsValid(string s)
    {
        Stack<char> stack = new Stack<char>();
        foreach (var c in s)
        {
            if (IsOpen(c))
            {
                stack.Push(c);
            }
            else
            {
                if (stack.Count > 0 && IsEqual(c, stack.Peek()))
                {
                    stack.Pop();
                }
                else
                {
                    return false;
                }
            }
        }

        return stack.Count == 0;
    }
    
    public bool IsOpen(char c)
    {
        return c == '(' || c == '[' || c == '{';
    }

    public bool IsEqual(char c, char top)
    {
        return ((c == ')' && top == '(') || (c == ']' && top == '[') || (c == '}' && top == '{'));
    }
}